include/swift/AST/ExtInfo.h

```diff
@@ -81,6 +83,26 @@ class ClangTypeInfo {
   void dump(llvm::raw_ostream &os, const clang::ASTContext &ctx) const;
 };
 
+// MARK: - UnexpectedClangTypeError
+/// Potential errors when trying to store a Clang type in an ExtInfo.
+struct UnexpectedClangTypeError {
+  enum class Kind {
+    NullForCOrBlock,
+    NonnullForNonCOrBlock,
+    NotBlockPointer,
+    NotFunctionPointerOrReference,
+    NonCanonical,
+  };
+  const Kind errorKind;
+  const clang::Type *type;
+
+  static Optional<UnexpectedClangTypeError> checkClangType(
+    SILFunctionTypeRepresentation fnRep, const clang::Type *type,
+    bool expectNonnullForCOrBlock, bool expectCanonical);
+
+  void dump();
+};
```

Comment:
This seems like a lot of abstraction for an assertion, but if you think it's necessary, we can roll with it.

Reply:
Yeah, I'm not a big fan of it, but it seemed like the cleanest approach given that we need to check for the error in three different places, including this one in `SILFunctionType::get(`.

```
  if (ctx.LangOpts.UseClangFunctionTypes) {
    if (auto error = ext.checkClangType()) {
      error.getValue().dump();
      llvm_unreachable("Unexpected Clang type in SILExtInfo.");
    }
  }
```